File: aip-otel-exporter/packages/python/src/aip_otel_exporter/attributes.py
"""
OpenTelemetry attribute name constants for AIP and AAP spans.

Primary namespace: aip.* and aap.* (vendor-specific).
Forward-compat aliases: gen_ai.evaluation.* for future OTel GenAI SIG alignment.
"""

# --- AIP Integrity Check Attributes ---

AIP_INTEGRITY_CHECKPOINT_ID = "aip.integrity.checkpoint_id"
AIP_INTEGRITY_VERDICT = "aip.integrity.verdict"
AIP_INTEGRITY_PROCEED = "aip.integrity.proceed"
AIP_INTEGRITY_RECOMMENDED_ACTION = "aip.integrity.recommended_action"
AIP_INTEGRITY_CONCERNS_COUNT = "aip.integrity.concerns_count"
AIP_INTEGRITY_AGENT_ID = "aip.integrity.agent_id"
AIP_INTEGRITY_CARD_ID = "aip.integrity.card_id"
AIP_INTEGRITY_SESSION_ID = "aip.integrity.session_id"
AIP_INTEGRITY_THINKING_HASH = "aip.integrity.thinking_hash"
AIP_INTEGRITY_ANALYSIS_MODEL = "aip.integrity.analysis_model"
AIP_INTEGRITY_ANALYSIS_DURATION_MS = "aip.integrity.analysis_duration_ms"
AIP_INTEGRITY_THINKING_TOKENS = "aip.integrity.thinking_tokens"
AIP_INTEGRITY_TRUNCATED = "aip.integrity.truncated"
AIP_INTEGRITY_EXTRACTION_CONFIDENCE = "aip.integrity.extraction_confidence"

# --- AIP Conscience Attributes ---

AIP_CONSCIENCE_CONSULTATION_DEPTH = "aip.conscience.consultation_depth"
AIP_CONSCIENCE_VALUES_CHECKED_COUNT = "aip.conscience.values_checked_count"
AIP_CONSCIENCE_CONFLICTS_COUNT = "aip.conscience.conflicts_count"

# --- AIP Window Attributes ---

AIP_WINDOW_SIZE = "aip.window.size"
AIP_WINDOW_INTEGRITY_RATIO = "aip.window.integrity_ratio"
AIP_WINDOW_DRIFT_ALERT_ACTIVE = "aip.window.drift_alert_active"

# --- GenAI SIG Forward-Compat Aliases ---

GEN_AI_EVALUATION_VERDICT = "gen_ai.evaluation.verdict"
GEN_AI_EVALUATION_SCORE = "gen_ai.evaluation.score"

# --- AAP Verification Attributes ---

AAP_VERIFICATION_RESULT = "aap.verification.result"
AAP_VERIFICATION_SIMILARITY_SCORE = "aap.verification.similarity_score"
AAP_VERIFICATION_VIOLATIONS_COUNT = "aap.verification.violations_count"
AAP_VERIFICATION_WARNINGS_COUNT = "aap.verification.warnings_count"
AAP_VERIFICATION_TRACE_ID = "aap.verification.trace_id"
AAP_VERIFICATION_CARD_ID = "aap.verification.card_id"
AAP_VERIFICATION_DURATION_MS = "aap.verification.duration_ms"
AAP_VERIFICATION_CHECKS_PERFORMED = "aap.verification.checks_performed"

# --- AAP Coherence Attributes ---

AAP_COHERENCE_COMPATIBLE = "aap.coherence.compatible"
AAP_COHERENCE_SCORE = "aap.coherence.score"
AAP_COHERENCE_PROCEED = "aap.coherence.proceed"
AAP_COHERENCE_MATCHED_COUNT = "aap.coherence.matched_count"
AAP_COHERENCE_CONFLICT_COUNT = "aap.coherence.conflict_count"

# --- AAP Drift Detection Attributes ---

AAP_DRIFT_ALERTS_COUNT = "aap.drift.alerts_count"
AAP_DRIFT_TRACES_ANALYZED = "aap.drift.traces_analyzed"

# --- AIP Drift Alert Attributes (for events) ---

AIP_DRIFT_ALERT_ID = "aip.drift.alert_id"
AIP_DRIFT_AGENT_ID = "aip.drift.agent_id"
AIP_DRIFT_SESSION_ID = "aip.drift.session_id"
AIP_DRIFT_INTEGRITY_SIMILARITY = "aip.drift.integrity_similarity"
AIP_DRIFT_SUSTAINED_CHECKS = "aip.drift.sustained_checks"
AIP_DRIFT_SEVERITY = "aip.drift.severity"
AIP_DRIFT_DIRECTION = "aip.drift.drift_direction"
AIP_DRIFT_MESSAGE = "aip.drift.message"

# --- Span Names ---

SPAN_AIP_INTEGRITY_CHECK = "aip.integrity_check"
SPAN_AAP_VERIFY_TRACE = "aap.verify_trace"
SPAN_AAP_CHECK_COHERENCE = "aap.check_coherence"
SPAN_AAP_DETECT_DRIFT = "aap.detect_drift"

# --- Event Names ---

EVENT_AIP_CONCERN = "aip.concern"
EVENT_AIP_DRIFT_ALERT = "aip.drift_alert"
EVENT_AAP_VIOLATION = "aap.violation"
EVENT_AAP_DRIFT_ALERT = "aap.drift_alert"

# --- Metric Names ---

METRIC_AIP_INTEGRITY_CHECKS_TOTAL = "aip.integrity_checks.total"
METRIC_AIP_INTEGRITY_CHECKS_BY_VERDICT = "aip.integrity_checks.by_verdict"
METRIC_AIP_CONCERNS_TOTAL = "aip.concerns.total"
METRIC_AIP_ANALYSIS_DURATION = "aip.analysis.duration_ms"
METRIC_AIP_WINDOW_INTEGRITY_RATIO = "aip.window.integrity_ratio"
METRIC_AIP_DRIFT_ALERTS_TOTAL = "aip.drift_alerts.total"
METRIC_AAP_VERIFICATIONS_TOTAL = "aap.verifications.total"
METRIC_AAP_VIOLATIONS_TOTAL = "aap.violations.total"
METRIC_AAP_VERIFICATION_DURATION = "aap.verification.duration_ms"
METRIC_AAP_COHERENCE_SCORE = "aap.coherence.score"

File: aip-otel-exporter/packages/typescript/src/attributes.ts
/**
 * OpenTelemetry attribute name constants for AIP and AAP spans.
 *
 * Primary namespace: aip.* and aap.* (vendor-specific).
 * Forward-compat aliases: gen_ai.evaluation.* for future OTel GenAI SIG alignment.
 */

// --- AIP Integrity Check Attributes ---

export const AIP_INTEGRITY_CHECKPOINT_ID = "aip.integrity.checkpoint_id";
export const AIP_INTEGRITY_VERDICT = "aip.integrity.verdict";
export const AIP_INTEGRITY_PROCEED = "aip.integrity.proceed";
export const AIP_INTEGRITY_RECOMMENDED_ACTION =
  "aip.integrity.recommended_action";
export const AIP_INTEGRITY_CONCERNS_COUNT = "aip.integrity.concerns_count";
export const AIP_INTEGRITY_AGENT_ID = "aip.integrity.agent_id";
export const AIP_INTEGRITY_CARD_ID = "aip.integrity.card_id";
export const AIP_INTEGRITY_SESSION_ID = "aip.integrity.session_id";
export const AIP_INTEGRITY_THINKING_HASH = "aip.integrity.thinking_hash";
export const AIP_INTEGRITY_ANALYSIS_MODEL = "aip.integrity.analysis_model";
export const AIP_INTEGRITY_ANALYSIS_DURATION_MS =
  "aip.integrity.analysis_duration_ms";
export const AIP_INTEGRITY_THINKING_TOKENS = "aip.integrity.thinking_tokens";
export const AIP_INTEGRITY_TRUNCATED = "aip.integrity.truncated";
export const AIP_INTEGRITY_EXTRACTION_CONFIDENCE =
  "aip.integrity.extraction_confidence";

// --- AIP Conscience Attributes ---

export const AIP_CONSCIENCE_CONSULTATION_DEPTH =
  "aip.conscience.consultation_depth";
export const AIP_CONSCIENCE_VALUES_CHECKED_COUNT =
  "aip.conscience.values_checked_count";
export const AIP_CONSCIENCE_CONFLICTS_COUNT = "aip.conscience.conflicts_count";

// --- AIP Attestation Attributes ---

export const AIP_ATTESTATION_INPUT_COMMITMENT =
  "aip.attestation.input_commitment";
export const AIP_ATTESTATION_CHAIN_HASH = "aip.attestation.chain_hash";
export const AIP_ATTESTATION_MERKLE_ROOT = "aip.attestation.merkle_root";
export const AIP_ATTESTATION_SIGNATURE_VERIFIED =
  "aip.attestation.signature_verified";
export const AIP_ATTESTATION_CERTIFICATE_ID = "aip.attestation.certificate_id";
export const AIP_ATTESTATION_ZK_PROVEN = "aip.attestation.zk_proven";
export const AIP_ATTESTATION_ZK_PROOF_TIME_MS = "aip.attestation.zk_proof_time_ms";

// --- AIP Window Attributes ---

export const AIP_WINDOW_SIZE = "aip.window.size";
export const AIP_WINDOW_INTEGRITY_RATIO = "aip.window.integrity_ratio";
export const AIP_WINDOW_DRIFT_ALERT_ACTIVE = "aip.window.drift_alert_active";

// --- GenAI SIG Forward-Compat Aliases ---

export const GEN_AI_EVALUATION_VERDICT = "gen_ai.evaluation.verdict";
export const GEN_AI_EVALUATION_SCORE = "gen_ai.evaluation.score";

// --- OTel GenAI SemConv: upstream provider attribution ---
//
// Identifies the upstream LLM provider whose response this span's data
// originated from (anthropic / openai / gemini). Follows OTel SemConv GenAI:
// https://opentelemetry.io/docs/specs/semconv/gen-ai/
//
// Distinguished from AIP_INTEGRITY_ANALYSIS_MODEL: that names the *verifier*
// model (Haiku 4.5 in Mnemom's case); GEN_AI_REQUEST_MODEL names the
// customer's *upstream* model that produced the response being analyzed.
// Together they enable per-provider rollups for SLI-P1 / P2 / P3 (see
// safe-house-hardening/slos.md §"Provider-specific").

export const GEN_AI_SYSTEM = "gen_ai.system";
export const GEN_AI_REQUEST_MODEL = "gen_ai.request.model";

// --- Mnemom span role ---
//
// Distinguishes customer-path emissions from verifier-internal and harness
// traffic. Critical for honest per-provider rollups — the AIP verifier
// (Haiku 4.5) emits aip.integrity_check spans too, but those spans tag the
// customer's upstream provider on GEN_AI_SYSTEM, not the verifier's. Filter
// on `mnemom.span.role = 'customer'` to keep verifier-internal and harness-
// generated traffic out of per-provider customer SLOs.
//
// Values: "customer" (default), "verifier", "harness".

export const MNEMOM_SPAN_ROLE = "mnemom.span.role";

// --- AAP Verification Attributes ---

export const AAP_VERIFICATION_RESULT = "aap.verification.result";
export const AAP_VERIFICATION_SIMILARITY_SCORE =
  "aap.verification.similarity_score";
export const AAP_VERIFICATION_VIOLATIONS_COUNT =
  "aap.verification.violations_count";
export const AAP_VERIFICATION_WARNINGS_COUNT =
  "aap.verification.warnings_count";
export const AAP_VERIFICATION_TRACE_ID = "aap.verification.trace_id";
export const AAP_VERIFICATION_CARD_ID = "aap.verification.card_id";
export const AAP_VERIFICATION_DURATION_MS = "aap.verification.duration_ms";
export const AAP_VERIFICATION_CHECKS_PERFORMED =
  "aap.verification.checks_performed";

// --- AAP Coherence Attributes ---

export const AAP_COHERENCE_COMPATIBLE = "aap.coherence.compatible";
export const AAP_COHERENCE_SCORE = "aap.coherence.score";
export const AAP_COHERENCE_PROCEED = "aap.coherence.proceed";
export const AAP_COHERENCE_MATCHED_COUNT = "aap.coherence.matched_count";
export const AAP_COHERENCE_CONFLICT_COUNT = "aap.coherence.conflict_count";

// --- AAP Drift Detection Attributes ---

export const AAP_DRIFT_ALERTS_COUNT = "aap.drift.alerts_count";
export const AAP_DRIFT_TRACES_ANALYZED = "aap.drift.traces_analyzed";

// --- AIP Drift Alert Attributes (for events) ---

export const AIP_DRIFT_ALERT_ID = "aip.drift.alert_id";
export const AIP_DRIFT_AGENT_ID = "aip.drift.agent_id";
export const AIP_DRIFT_SESSION_ID = "aip.drift.session_id";
export const AIP_DRIFT_INTEGRITY_SIMILARITY =
  "aip.drift.integrity_similarity";
export const AIP_DRIFT_SUSTAINED_CHECKS = "aip.drift.sustained_checks";
export const AIP_DRIFT_SEVERITY = "aip.drift.severity";
export const AIP_DRIFT_DIRECTION = "aip.drift.drift_direction";
export const AIP_DRIFT_MESSAGE = "aip.drift.message";

// --- Reclassification Attributes ---

export const RECLASSIFICATION_AGENT_ID =
  "gen_ai.safety.reclassification.agent_id";
export const RECLASSIFICATION_CHECKPOINT_ID =
  "gen_ai.safety.reclassification.checkpoint_id";
export const RECLASSIFICATION_TRACE_ID =
  "gen_ai.safety.reclassification.trace_id";
export const RECLASSIFICATION_BEFORE_VERDICT =
  "gen_ai.safety.reclassification.before_verdict";
export const RECLASSIFICATION_AFTER_CLASSIFICATION =
  "gen_ai.safety.reclassification.after_classification";
export const RECLASSIFICATION_REASON =
  "gen_ai.safety.reclassification.reason";
export const RECLASSIFICATION_SCORE_BEFORE =
  "gen_ai.safety.reclassification.score_before";
export const RECLASSIFICATION_SCORE_AFTER =
  "gen_ai.safety.reclassification.score_after";

// --- AIP Output Analysis Attributes ---

export const AIP_INTEGRITY_OUTPUT_HASH = "aip.integrity.output_hash";
export const AIP_INTEGRITY_OUTPUT_TOKENS = "aip.integrity.output_tokens";
export const AIP_INTEGRITY_OUTPUT_TRUNCATED = "aip.integrity.output_truncated";
export const AIP_INTEGRITY_ANALYSIS_SCOPE = "aip.integrity.analysis_scope";

// --- Policy Evaluation Attributes ---

export const POLICY_AGENT_ID = "policy.agent_id";
export const POLICY_POLICY_ID = "policy.policy_id";
export const POLICY_POLICY_VERSION = "policy.policy_version";
export const POLICY_VERDICT = "policy.verdict";
export const POLICY_VIOLATIONS_COUNT = "policy.violations_count";
export const POLICY_WARNINGS_COUNT = "policy.warnings_count";
export const POLICY_COVERAGE_PCT = "policy.coverage_pct";
export const POLICY_CONTEXT = "policy.context";
export const POLICY_DURATION_MS = "policy.duration_ms";
export const POLICY_ENFORCEMENT_MODE = "policy.enforcement_mode";

// --- Span Names ---

export const SPAN_AIP_INTEGRITY_CHECK = "aip.integrity_check";
export const SPAN_AAP_VERIFY_TRACE = "aap.verify_trace";
export const SPAN_AAP_CHECK_COHERENCE = "aap.check_coherence";
export const SPAN_AAP_DETECT_DRIFT = "aap.detect_drift";
export const SPAN_RECLASSIFICATION = "gen_ai.safety.reclassification";
export const SPAN_POLICY_EVALUATE = "policy.evaluate";

// --- Event Names ---

export const EVENT_AIP_CONCERN = "aip.concern";
export const EVENT_AIP_DRIFT_ALERT = "aip.drift_alert";
export const EVENT_AAP_VIOLATION = "aap.violation";
export const EVENT_AAP_DRIFT_ALERT = "aap.drift_alert";
export const EVENT_POLICY_VIOLATION = "policy.violation";

// --- Metric Names ---

export const METRIC_AIP_INTEGRITY_CHECKS_TOTAL =
  "aip.integrity_checks.total";
export const METRIC_AIP_INTEGRITY_CHECKS_BY_VERDICT =
  "aip.integrity_checks.by_verdict";
export const METRIC_AIP_CONCERNS_TOTAL = "aip.concerns.total";
export const METRIC_AIP_ANALYSIS_DURATION = "aip.analysis.duration_ms";
export const METRIC_AIP_WINDOW_INTEGRITY_RATIO =
  "aip.window.integrity_ratio";
export const METRIC_AIP_DRIFT_ALERTS_TOTAL = "aip.drift_alerts.total";
export const METRIC_AAP_VERIFICATIONS_TOTAL = "aap.verifications.total";
export const METRIC_AAP_VIOLATIONS_TOTAL = "aap.violations.total";
export const METRIC_AAP_VERIFICATION_DURATION =
  "aap.verification.duration_ms";
export const METRIC_AAP_COHERENCE_SCORE = "aap.coherence.score";

// --- Safe House Sideband Detection (T1-3.1 Piece 6 — ADR-040 + ADR-045 + ADR-047) ---
//
// Sideband sources are observer-side cron-driven detector firings that cross
// to runtime via pending_advisories. The four ratified sources are
// sideband.coherence, sideband.fault_line, sideband.fleet, and sideband.drift.
// Future axes extend additively per ADR-047 §1.

export const SAFE_HOUSE_SIDEBAND_SOURCE = "safe_house.sideband.source";
export const SAFE_HOUSE_SIDEBAND_AXIS = "safe_house.sideband.axis";
export const SAFE_HOUSE_SIDEBAND_TEAM_ID = "safe_house.sideband.team_id";
export const SAFE_HOUSE_SIDEBAND_FINDING_COUNT = "safe_house.sideband.finding_count";
export const SAFE_HOUSE_SIDEBAND_SEVERITY = "safe_house.sideband.severity";
export const SAFE_HOUSE_SIDEBAND_PATTERN_TYPE = "safe_house.sideband.pattern_type";

export const SPAN_SAFE_HOUSE_SIDEBAND_FINDING = "safe_house.sideband.finding";
export const EVENT_SAFE_HOUSE_SIDEBAND_FINDING = "safe_house.sideband.finding";
export const METRIC_SAFE_HOUSE_SIDEBAND_FINDINGS_TOTAL =
  "safe_house.sideband.findings.total";
